File: generate.js
#!/usr/bin/env node
// cli/generate.js

const path = require('path');
const fs = require('fs');
require('dotenv').config({ path: path.resolve(__dirname, '../.env') });

const { generateSlideContent } = require('../agents/content-agent');
const { buildPresentation } = require('../engines/js/engine');

function slugify(str) {
  return str.toLowerCase().replace(/[^a-z0-9]+/g, '-').replace(/^-|-$/g, '');
}

function parseArgs() {
  const args = process.argv.slice(2);
  const opts = { topic: '', slides: 8, engine: process.env.DEFAULT_ENGINE || 'js', template: 'business', output: process.env.OUTPUT_DIR || './output' };
  for (let i = 0; i < args.length; i++) {
    if (args[i] === '--topic') opts.topic = args[++i];
    if (args[i] === '--slides') opts.slides = parseInt(args[++i]);
    if (args[i] === '--engine') opts.engine = args[++i];
    if (args[i] === '--template') opts.template = args[++i];
    if (args[i] === '--output') opts.output = args[++i];
  }
  return opts;
}

async function main() {
  const opts = parseArgs();

  if (!opts.topic) {
    console.error('❌  Usage: node generate.js --topic "Your Topic" [--slides 8] [--engine js|python] [--template business]');
    process.exit(1);
  }

  if (!process.env.GEMINI_API_KEY) {
    console.error('❌  GEMINI_API_KEY not found in .env');
    process.exit(1);
  }

  console.log(`\n🤖 Generating content for: "${opts.topic}" (${opts.slides} slides)...`);

  let slideData;
  try {
    slideData = await generateSlideContent({ topic: opts.topic, slideCount: opts.slides, template: opts.template });
    console.log(`✅ Gemini returned ${slideData.slides.length} slides`);
  } catch (err) {
    console.error('❌ Content generation failed:', err.message);
    process.exit(1);
  }

  const filename = slugify(opts.topic);
  console.log(`🛠  Building presentation (engine: ${opts.engine})...`);

  try {
    const outPath = await buildPresentation({ slideData, outputDir: opts.output, filename });
    console.log(`\n✅ Done! Saved to: ${outPath}\n`);
  } catch (err) {
    console.error('❌ Build failed:', err.message);
    process.exit(1);
  }
}

main();

File: generate.py
#!/usr/bin/env python3
# cli/generate.py

import argparse
import os
import re
import sys

# Add project root to path
sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), '..')))

from dotenv import load_dotenv
load_dotenv(dotenv_path=os.path.join(os.path.dirname(__file__), '../.env'))

from agents.content_agent import generate_slide_content
from engines.python.engine import build_presentation


def slugify(text: str) -> str:
    text = text.lower()
    text = re.sub(r'[^a-z0-9]+', '-', text)
    return text.strip('-')


def main():
    parser = argparse.ArgumentParser(description='Gemini-powered PowerPoint generator (Python engine)')
    parser.add_argument('--topic', required=True, help='Presentation topic')
    parser.add_argument('--slides', type=int, default=8, help='Number of slides (default: 8)')
    parser.add_argument('--template', default='business', help='Template style: business | technical | minimal')
    parser.add_argument('--output', default=os.getenv('OUTPUT_DIR', './output'), help='Output directory')
    args = parser.parse_args()

    if not os.getenv('GEMINI_API_KEY'):
        print('❌  GEMINI_API_KEY not found in .env')
        sys.exit(1)

    print(f'\n🤖 Generating content for: "{args.topic}" ({args.slides} slides)...')

    try:
        slide_data = generate_slide_content(topic=args.topic, slide_count=args.slides, template=args.template)
        print(f'✅ Gemini returned {len(slide_data["slides"])} slides')
    except Exception as e:
        print(f'❌ Content generation failed: {e}')
        sys.exit(1)

    filename = slugify(args.topic)
    print(f'🛠  Building presentation (python-pptx engine)...')

    try:
        out_path = build_presentation(slide_data=slide_data, output_dir=args.output, filename=filename)
        print(f'\n✅ Done! Saved to: {out_path}\n')
    except Exception as e:
        print(f'❌ Build failed: {e}')
        sys.exit(1)


if __name__ == '__main__':
    main()
